keps/sig-api-machinery/4020-unknown-version-interoperability-proxy/README.md

```diff
@@ -0,0 +1,989 @@
+
+This proposal addresses both problems.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+1. **Network connectivity isues between apiservers**
+    
+    Cluster admins might not read the release notes and realize they should enable network/firewall connectivity between apiservers. In this case clients will receive 503s instead of transparently being proxied. 503 is still safer than today's behavior. We will clearly document the steps needed to enable the feature and also include steps to verify that the feature is working as intended. Looking at the following exposed metrics can help wth that 
+    1. `kubernetes_uvip_proxied_request_total` to monitor the number of UVIP-proxies requests. This metric can tell us the number of requests that were successfully proxied and the ones that failed
+    2. `apiserver_request_total` to check the success/error status of the requests
+
+2. **Increase in egress bandwidth**
+    
+    Requests will consume egress bandwidth for 2 apiservers when proxied. We can cap the number if needed, but upgrades aren't that frequent and few resources are changed on releases, so these requests should not be common. We will count them with a metric.
+
+3. **Increase in request traffic directed at destination kube-apiserver**
+
+    There could be a large volume of requests for a specific resource which might result in the identified apiserver being unable to serve the proxied requests. This scenario should not occur too frequently, since resource types which have large request volume should not be added or removed during an upgrade - that would cause other problems, too.
+
+4. **Indefinite rerouting of the request**
+    
+    We should ensure at most one proxy, rather than proxying the request over and over again (if the source apiserver has an incorrect understanding of what the destination apiserver can serve). To do this, we will add a new header such as `X-Kubernetes-UVIP-Rerouted:true` to the  request once it is determined that the request cannot be served by the local apiserver and should therefore be proxied.  
+    We will remove this header after the request is received by the destination apiserver (i.e. after the proxy has happened once) at which point it will be served locally.
+
+5. **Putting IP/endpoint and trust bundle control in user hands in REST APIs**
+    
+    To prevent server-side request forgeries we will not give control over information about apiserver IP/endpoint and the trust bundle (used to authenticate server while proxying) to users via REST APIs.
+
+## Design Details
+
+### Aggregation Layer
+
+![Alt text](https://user-images.githubusercontent.com/26771552/243512491-47c79882-1f32-4067-8e10-f8e16faa4997.png?raw=true "Optional Title")
```

Comment:
We should perform authz before proxying right?

Reply:
We are relying on the remote apiserver's authz flow to do that i.e. once the request is proxied to the peer. Will that not be sufficient?

Reply:
No, we pre-authorize requests before proxying them to aggregated APIs and we should do the same here.  I see no benefit of not pre-authorizing the request, especially since all you have to do is move the UIVP handler to the end of the chain.

Reply:
Got it, sure we can move the handler to the end of the chain, I dont have any arguments against that. Will try it out in my prototype and update the diagram after confirming that it works well.

Reply:
Updated diagram to move the UVIP handler after the Authz handler